File: src/main/java/name/gyger/jmoney/session/SessionService.java
package name.gyger.jmoney.session;

import name.gyger.jmoney.category.CategoryInitializer;
import org.springframework.context.event.ContextRefreshedEvent;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.Query;

@Service
@Transactional
public class SessionService {

    @PersistenceContext
    private EntityManager em;

    private final CategoryInitializer categoryInitializer;

    public SessionService(CategoryInitializer categoryInitializer) {
        this.categoryInitializer = categoryInitializer;
    }

    public Session getSession() {
        Query q = em.createQuery("SELECT s FROM Session s LEFT JOIN FETCH s.rootCategory LEFT JOIN FETCH s.splitCategory LEFT JOIN FETCH s.transferCategory");
        return (Session) q.getSingleResult();
    }

    public void initSession() {
        removeOldSession();
        Session session = new Session();
        categoryInitializer.initCategories(session);
        em.persist(session);
    }

    @EventListener
    public void handleContextRefresh(ContextRefreshedEvent event) {
        if (!isSessionAvailable()) {
            initSession();
        }
    }

    public void removeOldSession() {
        if (isSessionAvailable()) {
            Session oldSession = getSession();
            em.remove(oldSession);
        }
    }

    public boolean isSessionAvailable() {
        Query q = em.createQuery("SELECT COUNT(s) FROM Session s");
        return 1 == (Long) q.getSingleResult();
    }

}

File: src/main/kotlin/name/gyger/jmoney/session/SessionService.kt
package name.gyger.jmoney.session

import name.gyger.jmoney.category.CategoryFactory
import org.springframework.context.event.ContextRefreshedEvent
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class SessionService(private val sessionRepository: SessionRepository,
                     private val categoryFactory: CategoryFactory) {

    fun initSession() {
        removeOldSession()
        val rootCategory = categoryFactory.createRootCategory()
        val transferCategory = categoryFactory.createTransferCategory(rootCategory)
        val splitCategory = categoryFactory.createSplitCategory(rootCategory)
        val session = Session(rootCategory, transferCategory, splitCategory)
        categoryFactory.createNormalCategories(rootCategory)
        sessionRepository.save(session)
    }

    @EventListener
    fun handleContextRefresh(event: ContextRefreshedEvent?) {
        if (!isSessionAvailable()) {
            initSession()
        }
    }

    fun removeOldSession() {
        if (isSessionAvailable()) {
            sessionRepository.delete(sessionRepository.getSession())
        }
    }

    fun isSessionAvailable(): Boolean {
        return sessionRepository.isSessionAvailable()
    }

}
